### PR TITLE
New api method: get node config params

### LINKED ISSUE #1638: API for getting config
This would allow adding a node to the network by only providing it an IP address of a single existing node, and having it query for all other params.

### PR DESCRIPTION
**Pull request by avive**
## Motivation
Closes #1638 

Enable app to pull config params from a hosted deployment so it can start a node that joins a specific run.

## Changes
Mostly to the api module.

## Test Plan
Need to write tests.

## TODO
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

### CODE REVIEW

File: api/grpc_server.go [hunk 1]
```diff
@@ -240,20 +242,47 @@ func NewGrpcService(port int, net NetworkAPI, state StateAPI, tx TxAPI, txMempoo
...
+		PoETServer:             config.PoETServer,
+		SyncRequestTimeout:     uint32(config.SyncRequestTimeout),
+		AtxsPerBlock:           uint32(config.AtxsPerBlock),
+		Bootstrap:              config.P2P.SwarmConfig.Bootstrap,
+		BootstrapNodes:         config.P2P.SwarmConfig.BootstrapNodes,
+		Randcon:                uint32(config.P2P.SwarmConfig.RandomConnections),
+		MaxInbound:             uint32(config.P2P.MaxInboundPeers),
+		HareWakeupDelta:        uint32(config.HARE.WakeupDelta),
+		HareRoundDuration:      uint32(config.HARE.RoundDuration),
+		HareCommitteeSize:      0,
```
Comment (avive) [inline api/grpc_server.go]: @y0sher - can't find these configs - where are they?

Comment (avive) [reply api/grpc_server.go]: HareCommitteeSize and HareMaxAdversaries

**Comment by noamnelke** [comment]
I don't like this approach for two main reasons:

1. Every node config update (adding/removing a param) now requires a wallet update, since you're sending a ProtoBuf which the wallet has to interpret and convert to a config file / command line args. The two shouldn't be coupled. Moreover, it's trivial to export a `.toml` file of our config, which the wallet can just store in `~/spacemesh/` as-is for the node to pick up.
2. You're explicitly listing the network config params in the export mechanism. When a developer will add a network param in the future, they will not remember to update this mechanism and it will be hard to understand what went wrong. The alternative is to separate the config params into local and network config objects, so that a dev adding a param will be forced to declare if it's local or network related. You then take the network params as a whole and put them in a `net_config.toml` and serve that to nodes.

On a related note, I think it's critical that our testing framework uses this mechanism, or we'll only discover bugs in it when we release the buggy software, perhaps even the app test, to make discovering issues faster.

**Comment by avive** [comment]
Irrelevant as @noamnelke is working on an alternative solution to this issue.

**Comment by avive** [comment]
@noam - we need a solution for this for the f&f app.

**Comment by noamnelke** [comment]
@avive there's an open PR for this: https://github.com/spacemeshos/smops/pull/9

@jmacloue-ait has suggested to simplify it and I'm waiting for him to do that.

**Comment by avive** [comment]
Great. Do note that this issue blocks the app from going code complete in time for f&f milestone.

File: api/grpc_server.go [hunk 2]
```diff
@@ -240,20 +242,48 @@ func NewGrpcService(port int, net NetworkAPI, state StateAPI, tx TxAPI, txMempoo
...
```
Comment (avive) [inline api/grpc_server.go]: @y0sher - where can one find this config param?

**Comment by avive** [comment]
Here's a sample of the json api response for this method:
`{
		"LayerDurationSec":10,
		"LayersPerEpoch":3,
		"Hdist":5,
		"PoETServer":"spacemesh-testnet-poet-grpc-lb-949d0cde858743fb.elb.us-east-1.amazonaws.com:50002",
		"SyncRequestTimeout":2000,
		"AtxsPerBlock":100,
		"BootstrapNodes":["spacemesh://9n4AnWsnVgUBCvRZA9Ai55sh6iN2HQmLE7u1vuS6pby4@15.165.9.50:63163",
		"spacemesh://9n4AnWsnVgUBCvRZA9Ai55sh6iN2HQmLE7u1vuS6pby4@15.165.9.50:63161"],
		"Randcon":5,
		"MaxInbound":100,
		"HareWakeupDelta":10,
		"HareRoundDuration":2,
		"HareCommitteeSize":10,
		"HareMaxAdversaries":5,
		"EligibilityConf":25,
		"GenesisActiveSize":5,
		"PostSize":"1048576",
		"PostLabels":"5000000",
		"GenesisTime":"2020-01-20T20:18:23+02:00",
		"NetworkId":1
	}`

File: api/api_test.go [hunk 3]
```diff
@@ -2,34 +2,35 @@ package api
...
+	nodeconf "github.com/spacemeshos/go-spacemesh/config"
 	"github.com/spacemeshos/go-spacemesh/mesh"
 	"github.com/spacemeshos/go-spacemesh/miner"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
 	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
 	"github.com/spacemeshos/go-spacemesh/p2p/service"
 	"github.com/spacemeshos/go-spacemesh/priorityq"
 	"github.com/spacemeshos/go-spacemesh/signing"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/net/context"
```
Comment (y0sher) [inline api/api_test.go]: I know this was here before you but I'd replace it with "context" (newer versions of go moved the package from the x package)

another thing is it seems like you reorged the imports, I liked the previous way better then this..

File: api/api_test.go [hunk 4]
```diff
@@ -491,11 +575,11 @@ func launchServer(t *testing.T) func() {
 	return func() {
 		jsonService.StopService()
 		grpcService.StopService()
-	}
+	}, nodeConf
```
Comment (y0sher) [inline api/api_test.go:589]: why do you need to return this if it's always the default config from your function?

File: api/grpc_server.go [hunk 5]
```diff
@@ -240,20 +242,48 @@ func NewGrpcService(port int, net NetworkAPI, state StateAPI, tx TxAPI, txMempoo
 			time.Minute * 3,
 		}),
 	}
+
+	nodeConfigParams := &pb.NodeConfigParams{
```
Comment (y0sher) [inline api/grpc_server.go]: I would extract this to a function just to reduce the length of this one and for clarity. something like
`NodeConfigParamsFromConfig(config.Config) *pb.NodeConfigParams`

File: cmd/node/node.go [hunk 6]
```diff
@@ -883,8 +883,9 @@ func (app *SpacemeshApp) Start(cmd *cobra.Command, args []string) {
 	// start api servers
 	if apiConf.StartGrpcServer || apiConf.StartJSONServer {
 		// start grpc if specified or if json rpc specified
-		layerDuration := app.Config.LayerDurationSec
-		app.grpcAPIService = api.NewGrpcService(apiConf.GrpcServerPort, app.P2P, app.state, app.mesh, app.txPool, app.atxBuilder, app.oracle, app.clock, postClient, layerDuration, app)
+		//app.Config
```
Comment (y0sher) [inline cmd/node/node.go]: remove comments

File: crypto/vrf.go [hunk 7]
```diff
@@ -3,7 +3,7 @@ package crypto
 import (
 	"crypto/rand"
 	"errors"
-	"golang.org/x/crypto/ed25519"
+	"github.com/spacemeshos/ed25519"
```
Comment (y0sher) [inline crypto/vrf.go]: this isn't really related to this PR. I would appreciate if we can extract to a different pr and I'd like @barakshani to take a look. also I think @gavraz was involved so he might have good input.

**Comment by avive** [comment]
@y0sher - I reverted back the vrf ed25519 import - removed from this PR.

**Comment by avive** [comment]
bors try

File: api/grpc_server.go [hunk 8]
```diff
@@ -226,8 +228,37 @@ type TxAPI interface {
...
+		HareCommitteeSize:      uint32(config.HARE.N),
+		HareMaxAdversaries:     uint32(config.HARE.F),
+		EligibilityConf:        uint32(config.HareEligibility.ConfidenceParam),
+		EligibilityEpochOffset: uint32(config.HareEligibility.EpochOffset),
+		GenesisActiveSize:      uint32(config.GenesisActiveSet),
+		PostSize:               config.POST.SpacePerUnit,
+		PostLabels:             config.POST.LabelsLogRate,
+		GenesisTime:            config.GenesisTime,
+		NetworkId:              uint32(config.P2P.NetworkID),
+	}
```
Comment (y0sher) [inline api/grpc_server.go:258]: looks good in terms of the params
I would run a test with this branch to see if it achieves what you need before merging though

File: api/api_test.go (hunk 4 above)
Comment (y0sher) [reply api/api_test.go:589]: do didn't address this comment

**Comment by y0sher** [comment]
All looks good and you addressing everything thus I'm approving though in order to make sure we don't need to revisit this again soon please test that this feature helps to achieve what you needed before merging. good job ;)

**Review by y0sher — APPROVED** [review]

**Comment by avive** [comment]
We don't need this feature anymore as we are now using toml file from a web endpoint to configure new nodes. This might be useful for testing but def not for TNFF.
